js/buffer.ts

```diff
@@ -0,0 +1,219 @@
+ * methods. Based on https://golang.org/pkg/bytes/#Buffer
+ */
+export class Buffer implements Reader, Writer {
+  private buf: Uint8Array; // contents are the bytes buf[off : len(buf)]
+  private off = 0; // read at buf[off], write at buf[buf.byteLength]
+
+  constructor(ab?: ArrayBuffer) {
+    if (ab == null) {
+      this.buf = new Uint8Array(0);
+    } else {
+      this.buf = new Uint8Array(ab, 0, ab.byteLength);
+    }
+  }
+
+  /** bytes() returns a slice holding the unread portion of the buffer.
+   * The slice is valid for use only until the next buffer modification (that
+   * is, only until the next call to a method like read(), write(), reset(), or
+   * truncate()). The slice aliases the buffer content at least until the next
+   * buffer modification, so immediate changes to the slice will affect the
+   * result of future reads.
+   */
+  bytes(): Uint8Array {
+    return this.buf.subarray(this.off);
+  }
+
+  /** toString() returns the contents of the unread portion of the buffer
+   * as a string.
+   */
+  toString(): string {
+    const decoder = new TextDecoder();
+    return decoder.decode(this.buf.subarray(this.off));
+  }
+
+  /** empty() returns whether the unread portion of the buffer is empty. */
+  empty() {
+    return this.buf.byteLength <= this.off;
+  }
+
+  /** length is a getter that returns the number of bytes of the unread
+   * portion of the buffer
+   */
+  get length() {
+    return this.buf.byteLength - this.off;
+  }
+
+  /** Cap returns the capacity of the buffer's underlying byte slice, that is,
+   * the total space allocated for the buffer's data.
+   */
+  cap(): number {
+    return this.buf.buffer.byteLength;
+  }
+
+  /** truncate() discards all but the first n unread bytes from the buffer but
+   * continues to use the same allocated storage.  It throws if n is negative or
+   * greater than the length of the buffer.
+   */
+  truncate(n: number): void {
+    if (n === 0) {
+      this.reset();
+      return;
+    }
+    if (n < 0 || n > this.length) {
+      throw Error("bytes.Buffer: truncation out of range");
+    }
+    this._reslice(this.off + n);
+  }
+
+  /** reset() resets the buffer to be empty, but it retains the underlying
+   * storage for use by future writes. reset() is the same as truncate(0)
+   */
+  reset() {
+    this._reslice(0);
+    this.off = 0;
+  }
+
+  /** _tryGrowByReslice() is a inlineable version of grow for the fast-case
+   * where the internal buffer only needs to be resliced. It returns the index
+   * where bytes should be written and whether it succeeded.
+   */
+  private _tryGrowByReslice(n: number): [number, boolean] {
+    const l = this.buf.byteLength;
+    if (n <= this.cap() - l) {
+      this._reslice(l + n);
+      return [l, true];
+    }
+    return [0, false];
+  }
+
+  private _reslice(len: number): void {
+    assert(len <= this.buf.buffer.byteLength);
+    this.buf = new Uint8Array(this.buf.buffer, 0, len);
```

Comment:
reslice is called very often ... I don't think it's so great for performance to create a new TypedArray every time.
Why is that? Since there is already the `off` member, it'd be just as easy to also have a `len` or `end` member right?